src/nix/comment.cc

```diff
@@ -0,0 +1,224 @@
+//
+// This module does not support tab ('\t') characters. In some places
+// they are treated as single spaces. They should be avoided.
+
+namespace nix {
+
+// parseDoc will try to recover a Doc by looking at the text that leads up to a term
+// definition.
+struct Doc parseDoc(std::string sourcePrefix);
+
+// stripComment unpacks a comment, by unindenting and stripping " * " prefixes as
+// applicable. The argument should include any preceding whitespace.
+std::string stripComment(std::string rawComment);
+
+// Consistent unindenting. It will only remove entire columns.
+std::string unindent(std::string s);
+
+std::string trimUnindent(std::string s) {
+    return trim(unindent(s));
+}
+
+// SLOW, probably O(n^2)
+std::string stripPrefix(std::string prefix, std::string s) {
+    size_t n = std::min(prefix.length(), s.length());
+    size_t i;
+    for (i = 0; i < n; i++) {
+        if (s[i] != prefix[i])
+            return s;
+    }
+    return s.erase(0, i);
+}
+
+struct Doc lookupDoc(Pos & pos) {
+
+    try {
+        std::ifstream ifs(static_cast<const std::string>(pos.file));
+        std::stringstream buffer;
+
+        // read up to line and column with unix newlines
+
+        std::string line;
+        for (size_t i = 0;
+             getline(ifs, line) && i < pos.line-1;
+             i++) {
+            buffer << line << "\n";
+        }
+        buffer << line.substr(0, pos.column-1);
+
+        return parseDoc(buffer.str());
+    } catch (std::exception e) {
+        std::cout << "Caught exception: " << e.what() << std::endl;
+        return emptyDoc;
+    }
+}
+
+// See lambdas in parseDoc
+int countLambdas(std::string piece) {
+    return std::count(piece.begin(), piece.end(), ':');
+}
+
+// Try to recover a Doc by looking at the text that leads up to a term
+// definition
+struct Doc parseDoc(std::string sourcePrefix) {
+
+    std::string wss("[ \t\r\n]*");
+    std::string spaces("[ \t]*");
+
+    std::string singleLineComment(spaces + "#[^\r\n]*(?:\n|\r\n)");
+    std::string multiSingleLineComment("(?:" + singleLineComment + ")*");
+    std::string multiLineComment("\\/\\*(?:[^*]|\\*+[^*/])*\\*+\\/");
+    std::string commentUnit("(" + multiSingleLineComment + "|" + spaces + multiLineComment + ")" + wss);
+
+    std::string ident("[a-zA-Z_][a-zA-Z0-9_'-]*" + wss);
+    std::string identKeep("([a-zA-Z_][a-zA-Z0-9_'-]*)" + wss);
+
+    // lvalue for nested attrset construction, but not matching
+    // quoted identifiers or ${...} or comments inbetween etc
+    std::string simplePath("(?:" + wss + ident + "\\.)*" + identKeep);
+
+    std::string lambda(ident + wss + ":" + wss);
+
+    // see countLambdas
+    std::string lambdas("((:?" + lambda + ")*)");
+
+    std::string assign("=" + wss);
+
+    std::string re(commentUnit + simplePath + assign + lambdas + "$");
+    std::regex e(re);
+    std::smatch matches;
+    regex_search(sourcePrefix, matches, e);
+
+    std::stringstream buffer;
+    if (matches.length() < 3) {
```

Comment:
I really don't know what kind of standards the nix code base has in general.. i usually try to avoid "magic numbers" and would introduce some `#define` here or so.. same for the indices to access matches below. Just a thought.